server/src/utils/media.ts

```diff
@@ -979,12 +988,21 @@ export class RkmppHwDecodeConfig extends RkmppSwDecodeConfig {
   getFilterOptions(videoStream: VideoStreamInfo) {
     if (this.shouldToneMap(videoStream)) {
       const { primaries, transfer, matrix } = this.getColors();
-      return [
-        `scale_rkrga=${this.getScaling(videoStream)}:format=p010:afbc=1`,
-        'hwmap=derive_device=opencl:mode=read',
-        `tonemap_opencl=format=nv12:r=pc:p=${primaries}:t=${transfer}:m=${matrix}:tonemap=${this.config.tonemap}:desat=0:tonemap_mode=lum:peak=100`,
-        'hwmap=derive_device=rkmpp:mode=write:reverse=1',
-        'format=drm_prime',
+      if (this.hasMaliOpenCL) {
+        return [ // use RKMPP for scaling, OpenCL for tone mapping
+          `scale_rkrga=${this.getScaling(videoStream)}:format=p010:afbc=1`,
+          'hwmap=derive_device=opencl:mode=read',
+          `tonemap_opencl=format=nv12:r=pc:p=${primaries}:t=${transfer}:m=${matrix}:tonemap=${this.config.tonemap}:desat=0:tonemap_mode=lum:peak=100`,
+          'hwmap=derive_device=rkmpp:mode=write:reverse=1',
+          'format=drm_prime',
+        ];
+      }
+      return [ // use RKMPP for scaling, CPU for tone mapping
+        `scale_rkrga=${this.getScaling(videoStream)}:format=nv12`,
```

Comment:
Set the format to p010 here.

Reply:
Unfortunately, after some digging, I think my chip(RK3566)'s 2D graphics module doesn't support 10bit output, therefore `p010` format (and other 10-bit format like `nv15`) is not supported. a total sw decoding might be needed here at least for RK3566

Reply:
Could you share the logs? It needs to support 10-bit decoding to even get to `scale_rkrga` since the decoding happens before this filter.

Reply:
sure, here's the log

<details>
<summary>p010</summary>

```
root@178a1f2d2a09:/usr/src/app/upload/library# /usr/bin/ffmpeg \
> -hwaccel rkmpp -hwaccel_output_format drm_prime -noautorotate \
> -i hdr.mp4 \
> -y -c:a copy -movflags faststart -fps_mode passthrough \
> -map 0:0 -map 0:1 -strict unofficial -g 256 -v verbose \
> -vf "scale_rkrga=-2:720:format=p010,\
> hwdownload,\
> format=yuv420p10le,\
> tonemapx=tonemap=hable:desat=0:p=bt709:t=bt709:m=bt709:r=pc:peak=100:format=nv12,\
> hwupload" \
> -c:v h264_rkmpp -level 51 -rc_mode CQP -qp_init 23 output.mp4
ffmpeg version 7.0.2-Jellyfin Copyright (c) 2000-2024 the FFmpeg developers
  built with gcc 12 (Debian 12.2.0-14)
  configuration: --prefix=/usr/lib/jellyfin-ffmpeg --target-os=linux --extra-version=Jellyfin --disable-doc --disable-ffplay --disable-ptx-compression --disable-static --disable-libxcb --disable-sdl2 --disable-xlib --enable-lto=auto --enable-gpl --enable-version3 --enable-shared --enable-gmp --enable-gnutls --enable-chromaprint --enable-opencl --enable-libdrm --enable-libxml2 --enable-libass --enable-libfreetype --enable-libfribidi --enable-libfontconfig --enable-libharfbuzz --enable-libbluray --enable-libmp3lame --enable-libopus --enable-libtheora --enable-libvorbis --enable-libopenmpt --enable-libdav1d --enable-libsvtav1 --enable-libwebp --enable-libvpx --enable-libx264 --enable-libx265 --enable-libzvbi --enable-libzimg --enable-libfdk-aac --arch=arm64 --cross-prefix=/usr/bin/aarch64-linux-gnu- --toolchain=hardened --enable-cross-compile --enable-rkmpp --enable-rkrga
  libavutil      59.  8.100 / 59.  8.100
  libavcodec     61.  3.100 / 61.  3.100
  libavformat    61.  1.100 / 61.  1.100
  libavdevice    61.  1.100 / 61.  1.100
  libavfilter    10.  1.100 / 10.  1.100
  libswscale      8.  1.100 /  8.  1.100
  libswresample   5.  1.100 /  5.  1.100
  libpostproc    58.  1.100 / 58.  1.100
Routing option strict to both codec and muxer layer
Selecting decoder 'hevc_rkmpp' because of requested hwaccel method rkmpp
Input #0, mov,mp4,m4a,3gp,3g2,mj2, from 'hdr.mp4':
  Metadata:
    major_brand     : isom
    minor_version   : 512
    compatible_brands: isomiso2mp41
    encoder         : Lavf61.7.100
  Duration: 00:00:20.52, start: 0.000000, bitrate: 9795 kb/s
  Stream #0:0[0x1](und): Video: hevc (Main 10), 1 reference frame (hev1 / 0x31766568), yuv420p10le(tv, bt2020nc/unknown/unknown, left), 2160x3840 [SAR 1:1 DAR 9:16], 9600 kb/s, 30 fps, 30 tbr, 16k tbn (default)
      Metadata:
        handler_name    : VideoHandler
        vendor_id       : [0][0][0][0]
        encoder         : Lavc61.19.100 libx265
      Side data:
        Content Light Level Metadata, MaxCLL=2000, MaxFALL=100
  Stream #0:1[0x2](und): Audio: aac (LC) (mp4a / 0x6134706D), 48000 Hz, stereo, fltp, 194 kb/s (default)
      Metadata:
        handler_name    : SoundHandler
        vendor_id       : [0][0][0][0]
[out#0/mp4 @ 0xaaaad5f89880] Adding streams from explicit maps...
[vost#0:0/h264_rkmpp @ 0xaaaad5f47a10] Created video stream from input stream 0:0
[tonemapx @ 0xaaaad5f2d2f0] Using CPU capability: NEON
[hevc_rkmpp @ 0xaaaad5f41640] Picked up an existing RKMPP hardware device
[aost#0:1/copy @ 0xaaaad5f96850] Created audio stream from input stream 0:1
Stream mapping:
  Stream #0:0 -> #0:0 (hevc (hevc_rkmpp) -> h264 (h264_rkmpp))
  Stream #0:1 -> #0:1 (copy)
[vost#0:0/h264_rkmpp @ 0xaaaad5f47a10] Starting thread...
[vf#0:0 @ 0xaaaad5f3d600] Starting thread...
[vist#0:0/hevc @ 0xaaaad5f3e000] [dec:hevc_rkmpp @ 0xaaaad5f89d30] Starting thread...
[in#0/mov,mp4,m4a,3gp,3g2,mj2 @ 0xaaaad5f2a900] Starting thread...
Press [q] to stop, [?] for help
[hevc_rkmpp @ 0xaaaad5f41640] Noticed an info change
[hevc_rkmpp @ 0xaaaad5f41640] Configured with size: 2160x3840 | pix_fmt: drm_prime | sw_pix_fmt: nv15
[tonemapx @ 0xffff9c002d00] Using CPU capability: NEON
[graph 0 input from stream 0:0 @ 0xffff9c006840] w:2160 h:3840 pixfmt:drm_prime tb:1/16000 fr:30/1 sar:1/1 csp:bt2020nc range:tv
[auto_scale_0 @ 0xffff9c008970] w:iw h:ih flags:'' interl:0
[Parsed_tonemapx_3 @ 0xffff9c002c50] auto-inserting filter 'auto_scale_0' between the filter 'Parsed_format_2' and the filter 'Parsed_tonemapx_3'
rga_api version 1.10.1_[3]
[Parsed_scale_rkrga_0 @ 0xffff9c0025e0] 'p010le' is only supported by RGA3
[Parsed_scale_rkrga_0 @ 0xffff9c0025e0] Failed to configure output pad on Parsed_scale_rkrga_0
[vf#0:0 @ 0xaaaad5f3d600] Error reinitializing filters!
[vf#0:0 @ 0xaaaad5f3d600] Task finished with error code: -38 (Function not implemented)
[vost#0:0/h264_rkmpp @ 0xaaaad5f47a10] Encoder thread received EOF
[vost#0:0/h264_rkmpp @ 0xaaaad5f47a10] Could not open encoder before EOF
[vost#0:0/h264_rkmpp @ 0xaaaad5f47a10] Task finished with error code: -22 (Invalid argument)
[vost#0:0/h264_rkmpp @ 0xaaaad5f47a10] Terminating thread with return code -22 (Invalid argument)
[vf#0:0 @ 0xaaaad5f3d600] Terminating thread with return code -38 (Function not implemented)
[vist#0:0/hevc @ 0xaaaad5f3e000] [dec:hevc_rkmpp @ 0xaaaad5f89d30] Decoder returned EOF, finishing
[vist#0:0/hevc @ 0xaaaad5f3e000] [dec:hevc_rkmpp @ 0xaaaad5f89d30] Terminating thread with return code 0 (success)
[vist#0:0/hevc @ 0xaaaad5f3e000] All consumers of this stream are done
[in#0/mov,mp4,m4a,3gp,3g2,mj2 @ 0xaaaad5f2a900] EOF while reading input
[in#0/mov,mp4,m4a,3gp,3g2,mj2 @ 0xaaaad5f2a900] Terminating thread with return code 0 (success)
[out#0/mp4 @ 0xaaaad5f89880] Nothing was written into output file, because at least one of its streams received no packets.
frame=    0 fps=0.0 q=0.0 Lsize=       0KiB time=N/A bitrate=N/A speed=N/A    
[AVIOContext @ 0xaaaad5f973a0] Statistics: 0 bytes written, 0 seeks, 0 writeouts
[in#0/mov,mp4,m4a,3gp,3g2,mj2 @ 0xaaaad5f2a900] Input file #0 (hdr.mp4):
[in#0/mov,mp4,m4a,3gp,3g2,mj2 @ 0xaaaad5f2a900]   Input stream #0:0 (video): 17 packets read (762470 bytes); 2 frames decoded; 0 decode errors; 
[in#0/mov,mp4,m4a,3gp,3g2,mj2 @ 0xaaaad5f2a900]   Input stream #0:1 (audio): 961 packets read (497398 bytes); 
[in#0/mov,mp4,m4a,3gp,3g2,mj2 @ 0xaaaad5f2a900]   Total: 978 packets (1259868 bytes) demuxed
[AVIOContext @ 0xaaaad5f33020] Statistics: 25126543 bytes read, 0 seeks
Conversion failed!
```
</details>

it says `'p010le' is only supported by RGA3`

I don't know if it is the thing but according to [some documents](https://github.com/airockchip/librga/blob/fb5f019ea0191ec1c34f49ac8be447ac8921aadd/docs/Rockchip_Developer_Guide_RGA_EN.md#image-format-supported), RK3588 is the only chip that supports a 10 bit output. Other chip (like mine RK3566) can decode 10bit stream, but not output them.

and also, from [chip datasheet](https://www.boardcon.com/download/Rockchip_RK3566_Datasheet_V1.1.pdf):

![image](https://github.com/user-attachments/assets/8a99411b-f0b1-4050-a1b0-5e4ce56ac7f3)

Reply:
Hmm, is it possible to not set the format in `scale_rkrga` and just have `hwdownload,format=yuv420p10le` (or possibly `hwmap=mode=read,format= yuv420p10le`) after it?

Reply:
Tested these three:

```
/usr/bin/ffmpeg \
-hwaccel rkmpp -hwaccel_output_format drm_prime -noautorotate \
-i hdr.mp4 \
-y -c:a copy -movflags faststart -fps_mode passthrough \
-map 0:0 -map 0:1 -strict unofficial -g 256 -v verbose \
-vf "scale_rkrga=-2:720,\
hwdownload,\
format=yuv420p10le,\
tonemapx=tonemap=hable:desat=0:p=bt709:t=bt709:m=bt709:r=pc:peak=100:format=nv12,\
hwupload" \
-c:v h264_rkmpp -level 51 -rc_mode CQP -qp_init 23 output.mp4
```

```
/usr/bin/ffmpeg \
-hwaccel rkmpp -hwaccel_output_format drm_prime -noautorotate \
-i hdr.mp4 \
-y -c:a copy -movflags faststart -fps_mode passthrough \
-map 0:0 -map 0:1 -strict unofficial -g 256 -v verbose \
-vf "scale_rkrga=-2:720,\
hwmap=mode=read,\
hwdownload,\
format=yuv420p10le,\
tonemapx=tonemap=hable:desat=0:p=bt709:t=bt709:m=bt709:r=pc:peak=100:format=nv12,\
hwupload" \
-c:v h264_rkmpp -level 51 -rc_mode CQP -qp_init 23 output.mp4
```

```
/usr/bin/ffmpeg \
-hwaccel rkmpp -hwaccel_output_format drm_prime -noautorotate \
-i hdr.mp4 \
-y -c:a copy -movflags faststart -fps_mode passthrough \
-map 0:0 -map 0:1 -strict unofficial -g 256 -v verbose \
-vf "scale_rkrga=-2:720,\
hwmap=mode=read,\
format=yuv420p10le,\
tonemapx=tonemap=hable:desat=0:p=bt709:t=bt709:m=bt709:r=pc:peak=100:format=nv12,\
hwupload" \
-c:v h264_rkmpp -level 51 -rc_mode CQP -qp_init 23 output.mp4
```


all doesn't work. all report the same error: `'nv15' as output is not supported if RGA2 is requested`, where `nv15` is the original pixel format of the video. If no pixel format is defined, original video pixel format is used, but still, 10-bit output is not supported, so same error :(